core/shared/src/main/scala/zio/internal/FiberRuntime.scala

```diff
@@ -771,9 +764,24 @@ final class FiberRuntime[E, A](fiberId: FiberId.Runtime, fiberRefs0: FiberRefs,
   }
 
   @inline
-  private def popStackFrame(nextStackIndex: Int): Unit = {
-    _stack(nextStackIndex) = null // GC
```

Comment:
The more I think about this, the more I realise that it's a pretty dangerous thing.

What if instead we didn't GC when `nextStackIndex` is below X (perhaps 300 to coincide with trampolining)?

Reply:
I decided to go with an auto-gc threshold (of 128). Updated the PR description to match the new approach

Reply:
Even though I love the performance improvement, I think people will complain if we are holding onto (unnecessary) memory for arbitrary long periods of time.

One possibility is to clear out the entries when the run loop begins, basically by starting at stack index, and nulling until the first null.

This opens the door for making a null array, e.g. `val NullData = Array.fill[AnyRef](...)(null)` and then using the faster `arraycopy` to null out the extra entries.

However, we'd still have the problem of holding onto memory for an indeterminate amount of time.

How much does this single change contribute to the performance improvements?

Reply:
> One possibility is to clear out the entries when the run loop begins, basically by starting at stack index, and nulling until the first null.

If I'm understanding your recommendation correctly, I believe that this is similar to what's currently implemented, but instead clearing out the entries when the runloop starts we do it [when it exits](https://github.com/zio/zio/pull/8800/files#diff-9a664e551dbf1607e888df13c46b57913d52e59dd88d09b160cd238264961cb4R414) (this also means on every yield / async operation or termination).

Effectively we're holding on to objects _unnecessarily_ only while evaluating synchronous effects, which in almost all cases should be a very small period of time, and only for those objects above the `_stackIndex` until the first `null` entry. Once the runloop finishes, the amount of memory we're holding on to will be the same as previously.

> This opens the door for making a null array, e.g. val NullData = Array.fill[AnyRef](...)(null) and then using the faster arraycopy to null out the extra entries.

It's currently done iteratively but I like this recommendation better 👍 

> How much does this single change contribute to the performance improvements?

Between 5-10% of increase in throughput depending on the benchmark.